### PR TITLE
Bugfix: Make sure to use the keyProvider when adding an item to the list

### LINKED ISSUE #297: Using a custom ProvidesKey will cause an error in MultiChosenValueBox
When specifiyng a custom `ProvidesKey` the `MultiChosenValueBox` will throw a NPE in the [addItemToChosenListBox](https://github.com/ArcBees/gwtchosen/blob/db130c5ad811f86047b182b4bd509604454b310b/plugin/src/main/java/com/arcbees/chosen/client/gwt/MultipleChosenValueListBox.java#L160) function. 

The line should be changed to:  `int index = valueKeyToIndex.get(keyProvider.getKey(value));`

### PR DESCRIPTION
**Pull request by timeu**
Fixes #297

### CODE REVIEW

**Comment by meriouma** [comment]
Thanks for the contribution. Could you add a test please?

**Comment by olafleur** [comment]
LGTM

**Comment by timeu** [comment]
@meriouma: Should I add a new Testcase to the integration-test module ?

**Comment by meriouma** [comment]
Yes I think you'll need to create a new one for this case, then use it in ChosenIT. Just ping me if you need further help.
Thanks!

**Comment by timeu** [comment]
@meriouma: I added a testcase, however I don't know how to run the integration tests from th command line (tried `mvn test` in the command line). I only managed to run them manually by `mvn gwt:run -Pintegration-test`

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import com.google.gwt.text.shared.AbstractRenderer;
+import com.google.gwt.user.client.ui.RootPanel;
+import com.google.gwt.view.client.ProvidesKey;
+
+public class MultiValueListBoxWithCustomKeyProvider extends TestCase {
+
+    private final ChosenOptions options;
+    private MultipleChosenValueListBox<Car> listBox;
+
+    public static class CarKeyProvider implements ProvidesKey<Car> {
```
Comment (meriouma) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java]: Move the class definition before the fields

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Nullable
+    protected MultipleChosenValueListBox<Car> getListBox() {
+        return listBox;
+    }
+
+    protected List<Car> getAvailableCars() {
+        List<Car> cars = Lists.newArrayList();
+        for (int i = 0; i < CarBrand.values().length; i++) {
```
Comment (meriouma) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java]: Extract `CarBrand.values()` to a variable

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            cars.add(new Car(i,CarBrand.values()[i]));
```
Comment (meriouma) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java]: add space after `i,`

File: integration-test/src/test/java/ChosenIT.java [hunk 4]
```diff
@@ -323,6 +324,15 @@ public void testIsAccepted() {
         assertThat(getOptions()).isEqualTo(CarBrand.getAllNames(RENDERER));
     }
 
+    @Test
+    public void multiValueListBoxWithCustomProvider_setsItemsSelectedText() {
+        // Given
+        loadTestCase(new MultiValueListBoxWithCustomKeyProvider());
+
+        // Then
+        assertThat(getSelectedOptionText()).isEqualTo("2 items selected");
```
Comment (meriouma) [inline integration-test/src/test/java/ChosenIT.java]: This is only true for the MobileIT, not for DesktopIT. You should but the test in each IT class and they should have their own asserts

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    protected MultiValueListBoxWithCustomKeyProvider(ChosenOptions options) {
+        this.options = options;
+    }
+
+    @Override
+    public void run() {
+        listBox = new MultipleChosenValueListBox<Car>(new AbstractRenderer<Car>() {
+
```
Comment (meriouma) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java]: remove empty line

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/domain/Car.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.arcbees.chosen.integrationtest.client.domain;
+
+public class Car {
+
```
Comment (meriouma) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/domain/Car.java]: Remove empty line

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import com.google.gwt.user.client.ui.RootPanel;
+import com.google.gwt.view.client.ProvidesKey;
+
+public class MultiValueListBoxWithCustomKeyProvider extends TestCase {
+
+    private final ChosenOptions options;
+    private MultipleChosenValueListBox<Car> listBox;
+
+    public static class CarKeyProvider implements ProvidesKey<Car> {
+
```
Comment (meriouma) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java]: Remove empty line

File: integration-test/src/test/java/ChosenIT.java (hunk 4 above)
Comment (timeu) [inline integration-test/src/test/java/ChosenIT.java]: Is there a better way to test it? 
The bugfix only affects the `addItemToChosenListBox` function ? 
Also how can I run the tests ?

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/domain/Car.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2015 ArcBees Inc.
```
Comment (olafleur) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/domain/Car.java]: 2016

File: integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2015 ArcBees Inc.
```
Comment (olafleur) [inline integration-test/src/main/java/com/arcbees/chosen/integrationtest/client/testcases/MultiValueListBoxWithCustomKeyProvider.java]: 2016

**Comment by olafleur** [comment]
LGTM. @meriouma ?

**Pull request merged**